src/shims/unix/mem.rs

```diff
@@ -0,0 +1,207 @@
+        let new_size = this.read_scalar(new_size)?.to_machine_usize(this)?;
+        let flags = this.read_scalar(flags)?.to_i32()?;
+
+        // old_address must be a multiple of the page size
+        #[allow(clippy::integer_arithmetic)] // PAGE_SIZE is nonzero
+        if old_address.addr().bytes() % this.machine.page_size != 0 || new_size == 0 {
+            this.set_last_error(Scalar::from_i32(this.eval_libc_i32("EINVAL")))?;
+            return Ok(this.eval_libc("MAP_FAILED"));
+        }
+
+        if flags & this.eval_libc_i32("MREMAP_FIXED") != 0 {
+            throw_unsup_format!("Miri does not support mremap wth MREMAP_FIXED");
+        }
+
+        if flags & this.eval_libc_i32("MREMAP_DONTUNMAP") != 0 {
+            throw_unsup_format!("Miri does not support mremap wth MREMAP_DONTUNMAP");
+        }
+
+        if flags & this.eval_libc_i32("MREMAP_MAYMOVE") == 0 {
+            // We only support MREMAP_MAYMOVE, so not passing the flag is just a failure
+            this.set_last_error(Scalar::from_i32(this.eval_libc_i32("EINVAL")))?;
+            return Ok(Scalar::from_maybe_pointer(Pointer::null(), this));
+        }
+
+        let align = this.machine.page_align();
+        let ptr = this.reallocate_ptr(
+            old_address,
+            Some((Size::from_bytes(old_size), align)),
+            Size::from_bytes(new_size),
+            align,
+            MiriMemoryKind::Mmap.into(),
+        )?;
+        if let Some(increase) = new_size.checked_sub(old_size) {
+            // We just allocated this, the access is definitely in-bounds and fits into our address space.
+            // mmap guarantees new mappings are zero-init.
+            this.write_bytes_ptr(
+                ptr.offset(Size::from_bytes(old_size), this).unwrap().into(),
+                std::iter::repeat(0u8).take(usize::try_from(increase).unwrap()),
+            )
+            .unwrap();
+        }
+        // Memory mappings are always exposed
+        let (prov, _) = ptr.into_parts();
+        let Provenance::Concrete { alloc_id, tag } = prov else { unreachable!() };
+        intptrcast::GlobalStateInner::expose_ptr(this, alloc_id, tag)?;
+
+        Ok(Scalar::from_pointer(ptr, this))
+    }
+
+    fn munmap(
+        &mut self,
+        addr: &OpTy<'tcx, Provenance>,
+        length: &OpTy<'tcx, Provenance>,
+    ) -> InterpResult<'tcx, Scalar<Provenance>> {
+        let this = self.eval_context_mut();
+
+        let addr = this.read_pointer(addr)?;
+        let length = this.read_scalar(length)?.to_machine_usize(this)?;
+
+        // addr must be a multiple of the page size
+        #[allow(clippy::integer_arithmetic)] // PAGE_SIZE is nonzero
+        if addr.addr().bytes() % this.machine.page_size != 0 {
+            this.set_last_error(Scalar::from_i32(this.eval_libc_i32("EINVAL")))?;
+            // FIXME: The man page says this returns MAP_FAILED but that is type void*, and this
+            // function returns int.
```

Comment:
I assume it's `(int)MAP_FAILED` then?

Reply:
:shrug: I just re-read the Linux man page for this and it specifies not only that `munmap` returns -1 on error, but in the docs for `mmap`:

> the value MAP_FAILED (that  is,  (void *) -1)